src/main/kotlin/com/terraformation/backend/tracking/db/DeliveryStore.kt

```diff
@@ -0,0 +1,102 @@
+package com.terraformation.backend.tracking.db
+
+import com.terraformation.backend.auth.currentUser
+import com.terraformation.backend.customer.db.ParentStore
+import com.terraformation.backend.customer.model.requirePermissions
+import com.terraformation.backend.db.default_schema.SpeciesId
+import com.terraformation.backend.db.nursery.WithdrawalId
+import com.terraformation.backend.db.tracking.DeliveryId
+import com.terraformation.backend.db.tracking.PlantingSiteId
+import com.terraformation.backend.db.tracking.PlantingType
+import com.terraformation.backend.db.tracking.PlotId
+import com.terraformation.backend.db.tracking.tables.daos.DeliveriesDao
+import com.terraformation.backend.db.tracking.tables.daos.PlantingsDao
+import com.terraformation.backend.db.tracking.tables.pojos.DeliveriesRow
+import com.terraformation.backend.db.tracking.tables.pojos.PlantingsRow
+import com.terraformation.backend.db.tracking.tables.references.PLOTS
+import com.terraformation.backend.log.perClassLogger
+import com.terraformation.backend.nursery.db.WithdrawalNotFoundException
+import java.time.InstantSource
+import javax.annotation.ManagedBean
+import org.jooq.DSLContext
+import org.jooq.impl.DSL
+
+@ManagedBean
+class DeliveryStore(
+    private val clock: InstantSource,
+    private val deliveriesDao: DeliveriesDao,
+    private val dslContext: DSLContext,
+    private val parentStore: ParentStore,
+    private val plantingsDao: PlantingsDao,
+) {
+  private val log = perClassLogger()
+
+  fun createDelivery(
+      withdrawalId: WithdrawalId,
+      plantingSiteId: PlantingSiteId,
+      plotId: PlotId? = null,
+      quantities: Map<SpeciesId, Int>,
+  ): DeliveryId {
+    requirePermissions { createDelivery(plantingSiteId) }
+
+    val now = clock.instant()
+    val userId = currentUser().userId
+
+    if (plotId == null && plantingSiteHasPlots(plantingSiteId)) {
+      throw DeliveryMissingPlotException(plantingSiteId)
+    }
```

Comment:
So by this logic, we can deliver to a planting site without specifying a plot _only if_ the planting site does not have plots 👍

Reply:
If a `plotId` is specified that doesn't exist, does the database constraint block this from succeeding? I guess it should, right, because of the foreign key constraint on the plantings table?

Reply:
Correct, the foreign-key constraint will stop you from inserting a nonexistent plot ID. And with the additional constraints in the earlier PR in this series, you also can't insert a plot ID that belongs to a different planting site.